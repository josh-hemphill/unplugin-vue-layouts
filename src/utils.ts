import { normalizePath as normalizePathVite } from 'vite';

export function normalizePath(path: string) {
  path = path.startsWith('/') ? path : `/${path}`;
  return normalizePathVite(path);
}
