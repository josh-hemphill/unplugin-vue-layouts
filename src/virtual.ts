import { normalizePath } from './utils';
import type { Options } from './index';

export function createVirtualModuleID(name: string) {
  const virtualModuleId = `virtual:${name}`;
  const resolvedVirtualModuleId = `\0${virtualModuleId}`;
  return {
    virtualModuleId,
    resolvedVirtualModuleId,
  };
}

export async function createVirtualGlob(target: string, isSync: boolean) {
  return `import.meta.glob("${target}/**/*.vue", { eager: ${isSync} })`;
}

type VirtualModuleCodeOptions = Required<Options>;

export async function createVirtualModuleCode(options: VirtualModuleCodeOptions) {
  const { target, defaultLayout, importMode, skipTopLevelRouteLayout, wrapComponent } = options;
  const normalizedTarget = normalizePath(target);
  const isSync = importMode === 'sync';

  const skipCode = /* ts */ `
    // unplugin-vue-router adds a top-level route to the routing group, which we should skip (ref: https://github.com/JohnCampionJr/vite-plugin-vue-layouts/issues/134)
    /** @type {boolean} */
    const skipLayout = top && !route.component && route.children?.find(r => (r.path === '' || r.path === '/') && r.meta?.isLayout)
    if (skipLayout) {
      return route
    }
  `;

  function createLayoutRoute() {
    return /* ts */ `
    // Create a new route with the layout as the component
    /** @type {import('vue-router').RouteRecordRaw} */
    const layoutRoute = {
      path: route.path,
      component: layouts[layout],
      meta: { ...route.meta, isLayout: true },
      // Handle root path specially to avoid infinite nesting
      children: top && route.path === '/' 
        ? [route] 
        : [{ ...route, path: '', meta: { ...route.meta, isLayout: false } }]
    }
    return layoutRoute
    `;
  }

  function createLayoutComponent() {
    return /* ts */ `
    if (!route.component) {
      return route
    }
    /** @type {import('vue-router').RouteRecordRaw} */
    const wrappedRoute = {
      ...route,
      component: h('div', [
        h(layouts[layout]),
        h(define${isSync ? '' : 'Async'}Component(() => route.component())),
      ]),
      meta: {
        ...route.meta,
        isLayout: true
      }
    }
    return wrappedRoute
    `;
  }

  function fixIndentation(code: string, indent: number) {
    return code.replace(/\n/g, `\n${' '.repeat(indent)}`);
  }

  const importStatements = [wrapComponent ? `import { h, define${isSync ? '' : 'Async'}Component } from 'vue'` : '']
    .filter(Boolean)
    .join('\n');

  return /* ts */ `
    ${importStatements}

    /**
     * @param {Router} router
     * @param {boolean} [withLayout=false]
     * @returns {import('vue-router').RouteRecordRaw[] | (() => import('vue-router').RouteRecordRaw[])}
     */
    export function createGetRoutes(router, withLayout = false) {
      const routes = router.getRoutes()
      if (withLayout) {
        return routes
      }
      return () => routes.filter(route => !route.meta?.isLayout)
    }

    /**
     * @param {import('vue-router').RouteRecordRaw[]} routes
     * @returns {import('vue-router').RouteRecordRaw[]}
     */
    export function setupLayouts(routes) {
      /** @type {Record<string, any>} */
      const layouts = {}

      const modules = ${await createVirtualGlob(normalizedTarget, isSync)}

      Object.entries(modules).forEach(([name, module]) => {
        let key = name.replace("${normalizedTarget}/", '').replace('.vue', '')
        layouts[key] = ${isSync ? 'module.default' : 'module'}
      })

      /**
       * @param {import('vue-router').RouteRecordRaw[]} lRoutes
       * @param {boolean} [top=true]
       * @returns {import('vue-router').RouteRecordRaw[]}
       */
      function deepSetupLayout(lRoutes, top = true) {
        return lRoutes.map(route => {
          // Process children first
          if (route.children?.length > 0) {
            route.children = deepSetupLayout(route.children, false)
          }

          // Skip if already processed
          if (route.meta?.isLayout) {
            return route
          }

          const layout = route.meta?.layout ?? '${defaultLayout}'

          ${fixIndentation(skipTopLevelRouteLayout ? skipCode : '', 6)}
          
          if (layout && layouts[layout]) {
            ${fixIndentation(wrapComponent ? createLayoutComponent() : createLayoutRoute(), 8)}
          }

          return route
        })
      }

      return deepSetupLayout(routes)
    }
  `;
}
