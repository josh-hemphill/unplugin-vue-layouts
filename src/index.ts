/* eslint-disable n/prefer-global/process */
import type { Plugin } from 'vite';
import { createVirtualModuleCode, createVirtualModuleID } from './virtual';

export interface Options {
  /**
   * layouts dir
   * @default "src/layouts"
   */
  target?: string;
  /**
   * default layout
   * @default "default"
   */
  defaultLayout?: string;
  /**
   * default auto resolve
   */
  importMode?: 'sync' | 'async';
  /**
   * If set, wrap the route's internal component with a layout instead of adding a wrapping route
   * @default false
   */
  wrapComponent?: boolean;
  /**
   * If opened, fix:
   * https://github.com/JohnCampionJr/vite-plugin-vue-layouts/issues/134
   * @default false
   */
  skipTopLevelRouteLayout?: boolean;
}

export default function VueLayouts(options: Partial<Options> = {}): Plugin {
  const {
    target = 'src/layouts',
    defaultLayout = 'default',
    importMode = process.env.VITE_SSG ? 'sync' : 'async',
    skipTopLevelRouteLayout = false,
    wrapComponent = false,
  } = options;

  const { virtualModuleId, resolvedVirtualModuleId } = createVirtualModuleID('vue-layouts');

  return {
    name: 'unplugin-vue-layouts',
    resolveId(id) {
      if (id === virtualModuleId) {
        return resolvedVirtualModuleId;
      }
    },
    load(id) {
      if (id === resolvedVirtualModuleId) {
        return createVirtualModuleCode({
          target,
          importMode,
          defaultLayout,
          skipTopLevelRouteLayout,
          wrapComponent,
        });
      }
    },
  };
}
