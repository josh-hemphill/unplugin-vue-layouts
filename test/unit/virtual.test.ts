import { describe, expect, it } from 'vitest';
import { createVirtualGlob, createVirtualModuleCode, createVirtualModuleID } from '../../src/virtual';

describe('virtual module utilities', () => {
  it('creates virtual glob', async () => {
    const syncGlob = await createVirtualGlob('./layouts', true);
    const asyncGlob = await createVirtualGlob('./layouts', false);

    expect(syncGlob).toBe('import.meta.glob("./layouts/**/*.vue", { eager: true })');
    expect(asyncGlob).toBe('import.meta.glob("./layouts/**/*.vue", { eager: false })');
  });
});

describe('createVirtualModuleCode', () => {
  it('generates code with default layout', async () => {
    const code = await createVirtualModuleCode({
      target: './layouts',
      defaultLayout: 'default',
      importMode: 'sync',
      skipTopLevelRouteLayout: false,
      wrapComponent: false,
    });

    expect(code).toMatchSnapshot();
  });

  it('generates code with wrapped components', async () => {
    const code = await createVirtualModuleCode({
      target: './layouts',
      defaultLayout: 'default',
      importMode: 'async',
      skipTopLevelRouteLayout: false,
      wrapComponent: true,
    });

    expect(code).toMatchSnapshot();
  });

  it('generates code with top level route layout skipping', async () => {
    const code = await createVirtualModuleCode({
      target: './layouts',
      defaultLayout: 'default',
      importMode: 'sync',
      skipTopLevelRouteLayout: true,
      wrapComponent: false,
    });

    expect(code).toMatchSnapshot();
  });
});
