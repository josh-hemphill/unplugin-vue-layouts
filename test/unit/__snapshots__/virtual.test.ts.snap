// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createVirtualModuleCode > generates code with default layout 1`] = `
"
    

    /**
     * @param {Router} router
     * @param {boolean} [withLayout=false]
     * @returns {import('vue-router').RouteRecordRaw[] | (() => import('vue-router').RouteRecordRaw[])}
     */
    export function createGetRoutes(router, withLayout = false) {
      const routes = router.getRoutes()
      if (withLayout) {
        return routes
      }
      return () => routes.filter(route => !route.meta?.isLayout)
    }

    /**
     * @param {import('vue-router').RouteRecordRaw[]} routes
     * @returns {import('vue-router').RouteRecordRaw[]}
     */
    export function setupLayouts(routes) {
      /** @type {Record<string, any>} */
      const layouts = {}

      const modules = import.meta.glob("/layouts/**/*.vue", { eager: true })

      Object.entries(modules).forEach(([name, module]) => {
        let key = name.replace("/layouts/", '').replace('.vue', '')
        layouts[key] = module.default
      })

      /**
       * @param {import('vue-router').RouteRecordRaw[]} lRoutes
       * @param {boolean} [top=true]
       * @returns {import('vue-router').RouteRecordRaw[]}
       */
      function deepSetupLayout(lRoutes, top = true) {
        return lRoutes.map(route => {
          // Process children first
          if (route.children?.length > 0) {
            route.children = deepSetupLayout(route.children, false)
          }

          // Skip if already processed
          if (route.meta?.isLayout) {
            return route
          }

          const layout = route.meta?.layout ?? 'default'

          
          
          if (layout && layouts[layout]) {
            
            // Create a new route with the layout as the component
            /** @type {import('vue-router').RouteRecordRaw} */
            const layoutRoute = {
              path: route.path,
              component: layouts[layout],
              meta: { ...route.meta, isLayout: true },
              // Handle root path specially to avoid infinite nesting
              children: top && route.path === '/' 
                ? [route] 
                : [{ ...route, path: '', meta: { ...route.meta, isLayout: false } }]
            }
            return layoutRoute
            
          }

          return route
        })
      }

      return deepSetupLayout(routes)
    }
  "
`;

exports[`createVirtualModuleCode > generates code with top level route layout skipping 1`] = `
"
    

    /**
     * @param {Router} router
     * @param {boolean} [withLayout=false]
     * @returns {import('vue-router').RouteRecordRaw[] | (() => import('vue-router').RouteRecordRaw[])}
     */
    export function createGetRoutes(router, withLayout = false) {
      const routes = router.getRoutes()
      if (withLayout) {
        return routes
      }
      return () => routes.filter(route => !route.meta?.isLayout)
    }

    /**
     * @param {import('vue-router').RouteRecordRaw[]} routes
     * @returns {import('vue-router').RouteRecordRaw[]}
     */
    export function setupLayouts(routes) {
      /** @type {Record<string, any>} */
      const layouts = {}

      const modules = import.meta.glob("/layouts/**/*.vue", { eager: true })

      Object.entries(modules).forEach(([name, module]) => {
        let key = name.replace("/layouts/", '').replace('.vue', '')
        layouts[key] = module.default
      })

      /**
       * @param {import('vue-router').RouteRecordRaw[]} lRoutes
       * @param {boolean} [top=true]
       * @returns {import('vue-router').RouteRecordRaw[]}
       */
      function deepSetupLayout(lRoutes, top = true) {
        return lRoutes.map(route => {
          // Process children first
          if (route.children?.length > 0) {
            route.children = deepSetupLayout(route.children, false)
          }

          // Skip if already processed
          if (route.meta?.isLayout) {
            return route
          }

          const layout = route.meta?.layout ?? 'default'

          
          // unplugin-vue-router adds a top-level route to the routing group, which we should skip (ref: https://github.com/JohnCampionJr/vite-plugin-vue-layouts/issues/134)
          /** @type {boolean} */
          const skipLayout = top && !route.component && route.children?.find(r => (r.path === '' || r.path === '/') && r.meta?.isLayout)
          if (skipLayout) {
            return route
          }
        
          
          if (layout && layouts[layout]) {
            
            // Create a new route with the layout as the component
            /** @type {import('vue-router').RouteRecordRaw} */
            const layoutRoute = {
              path: route.path,
              component: layouts[layout],
              meta: { ...route.meta, isLayout: true },
              // Handle root path specially to avoid infinite nesting
              children: top && route.path === '/' 
                ? [route] 
                : [{ ...route, path: '', meta: { ...route.meta, isLayout: false } }]
            }
            return layoutRoute
            
          }

          return route
        })
      }

      return deepSetupLayout(routes)
    }
  "
`;

exports[`createVirtualModuleCode > generates code with wrapped components 1`] = `
"
    import { h, defineAsyncComponent } from 'vue'

    /**
     * @param {Router} router
     * @param {boolean} [withLayout=false]
     * @returns {import('vue-router').RouteRecordRaw[] | (() => import('vue-router').RouteRecordRaw[])}
     */
    export function createGetRoutes(router, withLayout = false) {
      const routes = router.getRoutes()
      if (withLayout) {
        return routes
      }
      return () => routes.filter(route => !route.meta?.isLayout)
    }

    /**
     * @param {import('vue-router').RouteRecordRaw[]} routes
     * @returns {import('vue-router').RouteRecordRaw[]}
     */
    export function setupLayouts(routes) {
      /** @type {Record<string, any>} */
      const layouts = {}

      const modules = import.meta.glob("/layouts/**/*.vue", { eager: false })

      Object.entries(modules).forEach(([name, module]) => {
        let key = name.replace("/layouts/", '').replace('.vue', '')
        layouts[key] = module
      })

      /**
       * @param {import('vue-router').RouteRecordRaw[]} lRoutes
       * @param {boolean} [top=true]
       * @returns {import('vue-router').RouteRecordRaw[]}
       */
      function deepSetupLayout(lRoutes, top = true) {
        return lRoutes.map(route => {
          // Process children first
          if (route.children?.length > 0) {
            route.children = deepSetupLayout(route.children, false)
          }

          // Skip if already processed
          if (route.meta?.isLayout) {
            return route
          }

          const layout = route.meta?.layout ?? 'default'

          
          
          if (layout && layouts[layout]) {
            
            if (!route.component) {
              return route
            }
            /** @type {import('vue-router').RouteRecordRaw} */
            const wrappedRoute = {
              ...route,
              component: h('div', [
                h(layouts[layout]),
                h(defineAsyncComponent(() => route.component())),
              ]),
              meta: {
                ...route.meta,
                isLayout: true
              }
            }
            return wrappedRoute
            
          }

          return route
        })
      }

      return deepSetupLayout(routes)
    }
  "
`;
